### PR TITLE
CI: split PR workflows, add paths

### PR DESCRIPTION
**Pull request by markgoddard**
- CI: Split .github/workflows/stackhpc-pull-request.yml into tox and aio workflows
- CI: Add paths to stackhpc-pull-request-tox-aio.yml workflow

This should reduce our CI overheads by only running aios when necessary. It will also make it easier to merge some changes.

### CODE REVIEW

**Comment by markgoddard** [comment]
Cancelled CI jobs because we can see they were triggered correctly.

File: .github/workflows/stackhpc-pull-request-aio.yml [hunk 1]
```diff
@@ -1,39 +1,48 @@
...
+      - 'etc/kayobe/ansible/growroot.yml'
+      - 'etc/kayobe/ansible/requirements.yml'
+      - 'etc/kayobe/ansible/scripts/aio-init.sh'
+      - '!etc/kayobe/bifrost.yml'
+      - '!etc/kayobe/bmc.yml'
+      - '!etc/kayobe/cephadm.yml'
+      - '!etc/kayobe/containers/**'
+      - '!etc/kayobe/environments/aufn-ceph/**'
+      - '!etc/kayobe/environments/ci-builder/**'
+      - '!etc/kayobe/environments/ci-multinode/**'
```
Comment (Alex-Welsh) [inline .github/workflows/stackhpc-pull-request-aio.yml]:
```suggestion
      - '!etc/kayobe/environments/**'
      - 'etc/kayobe/environments/ci-aio/**'
```

Comment (markgoddard) [reply .github/workflows/stackhpc-pull-request-aio.yml]: Like it

**Comment by markgoddard** [comment]
I suppose the difficulty is that we wouldn't be able to make the aio jobs required any more :(

Oh GH actions, you are no zuul.

**Comment by Alex-Welsh** [comment]
> I suppose the difficulty is that we wouldn't be able to make the aio jobs required any more :(
> 
> Oh GH actions, you are no zuul.

Can we fake running it and immediately return a success?

**Comment by markgoddard** [comment]
> > I suppose the difficulty is that we wouldn't be able to make the aio jobs required any more :(
> > Oh GH actions, you are no zuul.
> 
> Can we fake running it and immediately return a success?

Possibly, but it would be a bit misleading and would need something other than the `paths` selector

**Comment by Alex-Welsh** [comment]
How did the testing for this go? does it need more work?

**Review by mnasiadka — DISMISSED** [review]

**Comment by markgoddard** [comment]
> How did the testing for this go? does it need more work?

Looks like it doesn't accept the skipped jobs :(

**Comment by jovial** [comment]
> > How did the testing for this go? does it need more work?
> 
> Looks like it doesn't accept the skipped jobs :(

https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/collaborating-on-repositories-with-code-quality-features/troubleshooting-required-status-checks#handling-skipped-but-required-checks. Seems like the only option is skip all jobs within a workflow rather than the workflow itself :(

**Comment by markgoddard** [comment]
> > > How did the testing for this go? does it need more work?
> > 
> > 
> > Looks like it doesn't accept the skipped jobs :(
> 
> https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/collaborating-on-repositories-with-code-quality-features/troubleshooting-required-status-checks#handling-skipped-but-required-checks. Seems like the only option is skip all jobs within a workflow rather than the workflow itself :(

That's what it's doing already. I was relying on job dependencies though and have just tried adding the condition to every job. Could you approve again?

It does look like it's just not recognising the jobs though as it lists them as skipped and expected.

**Review by jovial — DISMISSED** [review]

**Comment by markgoddard** [comment]
I managed to tweak the branch protection rules to require the calling workflow's job rather than the called workflow, and this seems to now recognise the skipped results as success. I guess it never gets to the called workflow if the calling job is skipped.

Just need to confirm that this still requires the jobs to pass when they run :crossed_fingers:

**Comment by markgoddard** [comment]
> I managed to tweak the branch protection rules to require the calling workflow's job rather than the called workflow, and this seems to now recognise the skipped results as success. I guess it never gets to the called workflow if the calling job is skipped.
> 
> Just need to confirm that this still requires the jobs to pass when they run 🤞

Lame, it doesn't work:

image

I'll revert the required status checks change.

**Comment by markgoddard** [comment]
I tried pushing the condition into the called workflow, and this worked for the Build Kayobe image check, but not for the aio checks because they depend on the image job and so don't run.

**Comment by markgoddard** [comment]
Started a [discussion](https://github.com/orgs/community/discussions/72708) on this.

**Comment by markgoddard** [comment]
Broken build kayobe image job
No always()
Status not recognised, but not a problem since the build failed.

**Comment by markgoddard** [comment]
Changed file triggers jobs
Fixed build kayobe image job
No always()
Status recognised.

**Comment by markgoddard** [comment]
No changed files so jobs not triggered
No always()
Status not recognised.

**Comment by markgoddard** [comment]
No changed files so jobs not triggered
Using always()
Status recognised.

**Comment by markgoddard** [comment]
Changed file triggers jobs
Working build kayobe image job
Using always()
Status recognised.

**Comment by markgoddard** [comment]
Changed file triggers jobs
Failing build kayobe image job
Using always()
Status recognised.
aio jobs run but should be skipped

**Comment by markgoddard** [comment]
No changed files so jobs not triggered
Using ${{ ! failure() }}
Status recognised.

Using always() && needs.build-kayobe-image.result == 'success'
Status not recognised: result == 'skipped'.

**Comment by markgoddard** [comment]
No changed files so jobs not triggered
Using always() && needs.build-kayobe-image.result != 'failure'
Or using ${{ ! failure() }}
Status recognised.

**Comment by markgoddard** [comment]
Changed file triggers jobs
Failing build kayobe image job
Using always() && needs.build-kayobe-image.result != 'failure'
Or using ${{ ! failure() }}
aio jobs skipped & not recognised as success

**Comment by markgoddard** [comment]
Changed file triggers jobs
Working build kayobe image job
Using always() && needs.build-kayobe-image.result != 'failure'
Or using ${{ ! failure() }}
aio jobs run & recognised as success

**Comment by markgoddard** [comment]
Looks like either of these should do the trick

**Comment by markgoddard** [comment]
This PR is a bit of a mess. Closing in favour of https://github.com/stackhpc/stackhpc-kayobe-config/pull/726

**Pull request closed without merging**